### PR TITLE
Add a readme

### PR DESCRIPTION
**Pull request by errordeveloper**
It's long overdue! ~I know that more detailed usage examples would be quite good, but I'd leave that for the next PR.~

I'm really keen to hear feedback about the description, whether it's clear enough or a bit too "out there"... Please do keep in mind that this tool is relatively low-level, it's currently for the kind of users who'd otherwise run `flux push artifact` or possibly go DIY.

### CODE REVIEW

**Review by lnovara — COMMENTED** [review]
Following a call with @errordeveloper, I try to summarise my thoughts about Tape.

I really like the simplicity of the tool and how it helps in framing the configuration as data.

I would better highlight Tape's relation with tools such as Kustomize and Helm and also that, right now, adoption of Tape is limited to Flux users since AFAIK is the only GitOps tool that understands OCI artifacts.

Nice work! 🙌

**Comment by errordeveloper** [comment]
@lnovara thank you! I've added a paragraph about the ecosystem, please let me know if it's any good :)

File: README.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
+# Tape is for packaging applications
+
+Tape is a tool that allows software authors to package their entire applications as self-contained (taped) OCI images
+that can be deployed to a Kubernetes cluster. A taped OCI image contains all application components and configuration
+required to run all of the components together.
+
```
Comment (ipedrazas) [inline README.md:4]: We should add a "What's the problem?" section here

File: README.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+## Acknowledgements & Prior Art
+
+What Tape does is very much in the spirit of Docker images, but it extends the idea by shifting the perspective to configuration
+as an entry point to a map of dependencies, as opposed to forced separation of app images and configuration.
+
+It's not a novelty to package configuration in OCI, there exist many examples of this practice, but there is no interoperability.
+Tape's ambition is to commoditise the model and abstract configuration tooling, so that end-users don't need to think whether a
+particular app needs to be deployed with Helm, Kustomize or something else.
+
+Tape was directly inspired by `flux push artefact`. Incidentally, it also resembles some of the aspects of CNAB, but it is much
```
Comment (stefanprodan) [inline README.md]:
```suggestion
Tape was directly inspired by [flux push artifact](https://fluxcd.io/flux/cheatsheets/oci-artifacts/). Incidentally, it also resembles some of the aspects of CNAB, but it is much
```

**Review by stefanprodan — COMMENTED** [review]
I would replace `artefact` with `artifact`, as this is the spelling used by OCI, Flux, etc.

File: README.md (hunk 1 above)
Comment (errordeveloper) [reply README.md:4]: I've rewritten the intro, PTAL :)

**Comment by errordeveloper** [comment]
@stefanprodan @slimslenderslacks both of you mentioned that showing OCI JSON manifests would be helpful, I've added an example that shows all that there is. PTAL, whenever you have a chance =)

It's quite verbose, maybe I can use `<details/>` for some of these JSON objects, but the readme is a bit wordy overall anyway, I wish it to be shorter. Perhaps I can worry about that later.

File: README.md [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+              "isDir": true,
+              "git": {
+                "objectHash": "e5f73cd48e13a37c7f7c7b116d7da41e9adf7fd6",
+                "remotes": {
+                  "origin": [
+                    "https://github.com/stefanprodan/podinfo"
+                  ]
+                },
+                "reference": {
+                  "name": "refs/heads/master",
```
Comment (stefanprodan) [inline README.md]: At the top there is `git switch 6.4.1` but in here in shows `master`. I suggest to redo examples for the Git tag.

Comment (errordeveloper) [reply README.md]: `git switch 6.4.1` actually fails, have to use `git switch --detach 6.4.1`... 🙃

Comment (errordeveloper) [reply README.md]: Well spotted! I've fixed the output and opened https://github.com/docker/labs-brown-tape/issues/13 to address tags in attestations.

**Review by stefanprodan — APPROVED** [review]
LGTM 💯

**Review by ipedrazas — APPROVED** [review]
lgtm

File: README.md [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+## What is Tape?
+
+Tape is a tool that can package an entire application as a self-contained (taped) OCI image that can be deployed to a
+Kubernetes cluster. A taped OCI image contains all application components and Kubernetes resources required to run all
+of the components together.
+
+## What problem does it solve?
+
+The process of building and deploying an application to Kubernetes is highly fragmented. There are many choices that
+one has to make in implementing CI/CD. Leaving aside what CI vendor you pick, the starting point is the overall design of
```
Comment (kipz) [inline README.md]:
```suggestion
one has to make in implementing CI/CD. Leaving aside what CI vendor you pick, the starting point is the overall design of
```
```suggestion
one has to make when implementing CI/CD. Leaving aside what CI vendor you pick, the starting point is the overall design of
```

File: README.md [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+a pipeline, repository structure, Dockerfiles, use of OCI registry, mapping of revisions to image tags, rules of deployment
+based on tags, promotion between environments, as well as various kinds of ways to manage Kubernetes configs and how these
+configs get deployed to any particular cluster.
+
+Of course, there is some essential complexity, all CI pipelines will be different and everyone cannot use one and the same
+tool for philosophical reasons. However, as most CI systems rest on shell scripts, it's quite challenging to define
```
Comment (kipz) [inline README.md]:
```suggestion
tool for philosophical reasons. However, as most CI systems rely on shell scripts, it's quite challenging to define
```

File: README.md [hunk 6]
```diff
@@ -0,0 +1,462 @@
...
+contracts. What Tape does is all about having a simple artifact, which is a concrete contractual term, and Tape ensures
+it has certain properties, such as it references all formal dependencies by digest, i.e. dependencies that are OCI images
```
Comment (kipz) [inline README.md]:
```suggestion
it has certain properties, for example, it references all formal dependencies by digest, i.e. dependencies that are OCI images
```

File: README.md [hunk 7]
```diff
@@ -0,0 +1,462 @@
...
+referenced from a canonical `image` field in a Kubernetes resource (e.g. `spec.containers[].image`).
+
+Another issue that arises due to fragmentation is about collaboration between organisations that have made different choices,
```
Comment (kipz) [inline README.md:23]:
```suggestion
Another issue that arises due to fragmentation is about collaboration between organizations that have made different choices,
```

File: README.md [hunk 8]
```diff
@@ -0,0 +1,462 @@
...
+despite all using OCI images and Kubernetes APIs.  To illustrate this, let's ask ourselves - "Why should there be any difference
```
Comment (kipz) [inline README.md]:
```suggestion
despite all using OCI images and Kubernetes APIs.  To illustrate this, let's ask ourselves: "Why should there be any difference
```

File: README.md [hunk 9]
```diff
@@ -0,0 +1,462 @@
...
+in the mechanics of how my application is deployed to a Kubernetes cluster vs. someone else's application being deployed to the
+same cluster?". Of course, it all comes down to Kubernetes API resources, but there is often a lot that happens in between.
+Sometimes Helm is used, sometimes it's Kustomize, plain manifests are used with some bespoke automation around configs Git,
```
Comment (kipz) [inline README.md]:
```suggestion
Sometimes Helm is used, sometimes it's Kustomize, sometimes plain manifests are used with some bespoke automation around configs Git,
```

File: README.md [hunk 10]
```diff
@@ -0,0 +1,462 @@
...
+there many examples of configs defined scripting languages as well. Of course, it's very often a mix of a few approaches.
```
Comment (kipz) [inline README.md]:
```suggestion
and there many examples of configs defined in scripting languages as well. Of course, it's very often a mix of a few approaches.
```

File: README.md [hunk 11]
```diff
@@ -0,0 +1,462 @@
...
+transferable, and there is little to say about the benefits of having something special.
+
+Tape addresses the complexity by using OCI for the distribution of runnable code as well configuration required to run it
+correctly on Kubernetes. Tape introduces a notion of application artifact without introducing any particular model of how
+components are composed into an artifact. This model aims to be a layer of interoperability between different tools and
+provide a logical supply chain entry point and location for storing metadata.
+
+The best analogy is flatpack furniture. Presently, deployment of an application is as if flatpack hasn't been invented, so
+when someone orders a wooden cabinet, all that arrives in a box is just the pieces of wood, they have to shop for nuts,
+bolts, and tools. Of course, that might be desirable for some, as they have a well-stocked workshop with the best tools and
```
Comment (kipz) [inline README.md]:
```suggestion
bolts, and tools. Of course, that might be desirable for some, as they have a well stocked workshop with the best tools and
```

File: README.md [hunk 12]
```diff
@@ -0,0 +1,462 @@
...
+To summarise, Tape reduces the complexity of application delivery by packaging the entire application as an OCI image, providing
+a transferable artifact that includes config and all of the components, analogous to flatpack furniture. This notion of artifact
+is very important because it helps to define a concise contract.
+Tape also produces attestations about the provenance of the configuration source as well as any transforms it applies to the
+source. The attestations are attached to the resulting OCI image, so it helps with security and observability as well.
+
+## How does Tape work?
+
+> NB: This describes the current implementation that is very minimal and doesn't achieve all of the ambitious goals
+> as described above, namely it works only with plain YAML and JSON manifests and doesn't yet have key integrations,
```
Comment (kipz) [inline README.md]:
```suggestion
> as described above. Namely it works only with plain YAML and JSON manifests and doesn't yet have key integrations,
```

File: README.md [hunk 13]
```diff
@@ -0,0 +1,462 @@
...
+For each of the images, Tape searches of all well-known related tags, such as external signatures, attestations and
+SBOMs. Tape will make a copy of every application image and any tags related to it to a registry the user has specified.
+Once images are copied, it updates manifests with new references and bundles the result in an OCI artifact pushed to
+the same repo in the registry.
+
+Copying of all application images and referencing by digest is performed to ensure the application and its configuration
+are tightly coupled together to provide a single link in the supply chain as well as a single point of distribution
+and access control for the whole application.
+
+Tape also checks the VCS provenance of manifests, so if any manifest files are checked in Git, Tape will attest to what
```
Comment (kipz) [inline README.md]:
```suggestion
Tape also checks the VCS provenance of manifests, so if any manifest files are checked in to Git, Tape will attest to what
```

File: README.md [hunk 14]
```diff
@@ -0,0 +1,462 @@
...
+Git repository each file came from, all of the revision metadata, and whether it's been modified or not.
+Additionally, Tape attests to all key steps that it performs, e.g. original image references it detects and manifest
+checksums. It stores the attestations using in-toto format in an OCI artifact, the
```
Comment (kipz) [inline README.md]:
```suggestion
checksums. It stores the attestations using in-toto format in an OCI artifact.
```

File: README.md [hunk 15]
```diff
@@ -0,0 +1,462 @@
...
+either loosely coupled collections of resouces (like Kustomize), or opinionated application package formats (most
+notably Helm). One of the goals of Tape is to abstract the use of any tools that already exist while paving the way
+for innovation. Tape will attempt to integrate with most of the popular tools, and enable anyone to deploy applications
+from taped images without having to know if under the hood it will use Kustomize, Helm, just plain manifest, or something
+else entirely. The other goal is that users won't need to know about Tape either, perhaps someday `kubectl apply` could
+support OCI artifacts and there could be different ways of building the artifacts.
+
+### What kind of applications can Tape package?
+
+Tape doesn't infer an opinion of how the application is structured, namely what it consists of or doesn't consist of. It doesn't
```
Comment (kipz) [inline README.md]:
```suggestion
Tape doesn't infer an opinion of how the application is structured, what it consists of or doesn't consist of. It doesn't
```

**Review by kipz — COMMENTED** [review]
Minor language suggestions. Feel free to ignore. Looks good to me :)
